File: backend/arrlink/arr/factory.py
"""Adapter factory (Radarr + Sonarr)."""

from __future__ import annotations

from .base import BaseAdapter
from .radarr import RadarrAdapter
from .sonarr import SonarrAdapter

_ADAPTERS: dict[str, type[BaseAdapter]] = {
    RadarrAdapter.app_type: RadarrAdapter,
    SonarrAdapter.app_type: SonarrAdapter,
}


def get_adapter(app_type: str, url: str, api_key: str, timeout: float = 15.0) -> BaseAdapter:
    cls = _ADAPTERS.get(app_type)
    if cls is None:
        raise ValueError(f"unsupported app type: {app_type}")
    return cls(url, api_key, timeout=timeout)

File: backend-ts/src/arr/factory.ts
import type { BaseAdapter } from './base.js'
import { RadarrAdapter } from './radarr.js'
import { SonarrAdapter } from './sonarr.js'

const DEFAULT_TIMEOUT_MS = 15_000

export function getAdapter(
  appType: string,
  url: string,
  apiKey: string,
  timeoutMs: number = DEFAULT_TIMEOUT_MS,
): BaseAdapter {
  switch (appType) {
    case 'radarr':
      return new RadarrAdapter(url, apiKey, timeoutMs)
    case 'sonarr':
      return new SonarrAdapter(url, apiKey, timeoutMs)
    default:
      throw new Error(`unsupported app type: ${appType}`)
  }
}
